Nowdays CI frequently failing on API level 24, and 33

**Describe the bug**
* The unit coverage test case is failing nowadays frequently on the `SaveSearchToRecentsTest` https://github.com/kiwix/kiwix-android/actions/runs/7288588337/job/19861444001?pr=3624#step:8:516.
* On Android 13 the flaky test is `ZimHostFragment` which frequently fails.

![Screenshot from 2023-12-22 12-00-32](https://github.com/kiwix/kiwix-android/assets/34593983/ab3e4058-c9a5-410e-9e59-c12f978d63be)



**Expected behavior**

It should pass.

**Environment**
- Version of Kiwix Android: 3.9.1
- Device : CI
- OS version : Android 24, Android 33

**Logs**

```kotlin
2023-12-21T15:10:58.9969470Z [33m12-21 15:10:58.264 28792 28865 W System.err: testZimHostFragment(org.kiwix.kiwixmobile.webserver.ZimHostFragmentTest): run 3 failed.[0m
2023-12-21T15:10:58.9971710Z [33m12-21 15:10:58.265 28792 28865 W System.err: testZimHostFragment(org.kiwix.kiwixmobile.webserver.ZimHostFragmentTest): Giving up after 3 failures.[0m
2023-12-21T15:10:59.0008360Z [31m12-21 15:10:58.267 28792 28865 E TestRunner: failed: testZimHostFragment(org.kiwix.kiwixmobile.webserver.ZimHostFragmentTest)[0m
2023-12-21T15:10:59.0272060Z [31m12-21 15:10:58.269 28792 28865 E TestRunner: ----- begin exception -----[0m
2023-12-21T15:10:59.0274620Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: java.lang.RuntimeException: No view found with text WiFi connection detected[0m
2023-12-21T15:10:59.0276310Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: -[DecorView] visibility:visible[0m
2023-12-21T15:10:59.0277840Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: --[LinearLayout] visibility:visible[0m
2023-12-21T15:10:59.0279460Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: ---[ViewStub] id:android:id/action_mode_bar_stub visibility:gone[0m
2023-12-21T15:10:59.0281050Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: ---[FrameLayout] visibility:visible[0m
2023-12-21T15:10:59.0282870Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: ----[FitWindowsFrameLayout] id:org.kiwix.kiwixmobile:id/action_bar_root visibility:visible[0m
2023-12-21T15:10:59.0284760Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: -----[ContentFrameLayout] id:android:id/content visibility:visible[0m
2023-12-21T15:10:59.0286860Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: ------[DrawerLayout] id:org.kiwix.kiwixmobile:id/navigation_container contDesc:Open Drawer visibility:visible[0m
2023-12-21T15:10:59.0288910Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: -------[NestedCoordinatorLayout] visibility:visible[0m
2023-12-21T15:10:59.0290820Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: --------[FragmentContainerView] id:org.kiwix.kiwixmobile:id/nav_host_fragment visibility:visible[0m
2023-12-21T15:10:59.0293330Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: ---------[FragmentContainerView] id:org.kiwix.kiwixmobile:id/nav_host_fragment visibility:visible[0m
2023-12-21T15:10:59.0295260Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: ----------[ConstraintLayout] visibility:visible[0m
2023-12-21T15:10:59.0297020Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: -----------[Toolbar] id:org.kiwix.kiwixmobile:id/toolbar visibility:visible[0m
2023-12-21T15:10:59.0300590Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: ------------[ConstraintLayout] id:org.kiwix.kiwixmobile:id/toolbarWithSearchPlaceholder visibility:gone[0m
2023-12-21T15:10:59.0303830Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: -------------[MaterialTextView] visibility:visible[0m
2023-12-21T15:10:59.0305520Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: ------------[AppCompatTextView] text:WiFi Hotspot visibility:visible[0m
2023-12-21T15:10:59.0307310Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: ------------[AppCompatImageButton] contDesc:Open Drawer visibility:visible[0m
2023-12-21T15:10:59.0309010Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: ------------[ActionMenuView] visibility:visible[0m
2023-12-21T15:10:59.0311200Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: -----------[MaterialTextView] id:org.kiwix.kiwixmobile:id/serverTextView text:Select the files you wish to host on the server visibility:visible[0m
2023-12-21T15:10:59.0313870Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: -----------[AppCompatImageView] id:org.kiwix.kiwixmobile:id/shareServerUrlIcon contDesc:Share URL via other applications visibility:gone[0m
2023-12-21T15:10:59.0316200Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: -----------[RecyclerView] id:org.kiwix.kiwixmobile:id/recyclerViewZimHost visibility:visible[0m
2023-12-21T15:10:59.0318010Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: ------------[ConstraintLayout] visibility:visible[0m
2023-12-21T15:10:59.0319940Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: -------------[MaterialTextView] id:org.kiwix.kiwixmobile:id/header_language text:English visibility:visible[0m
2023-12-21T15:10:59.0322440Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: ------------[MaterialCardView] visibility:visible[0m
2023-12-21T15:10:59.0324370Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: -------------[ConstraintLayout] visibility:visible[0m
2023-12-21T15:10:59.0326180Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: --------------[Guideline] id:org.kiwix.kiwixmobile:id/horizontal_padding visibility:gone[0m
2023-12-21T15:10:59.0328650Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: --------------[Guideline] id:org.kiwix.kiwixmobile:id/vertical_padding visibility:gone[0m
2023-12-21T15:10:59.0330780Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: --------------[MaterialCheckBox] id:org.kiwix.kiwixmobile:id/itemBookCheckbox visibility:visible[0m
2023-12-21T15:10:59.0332980Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: --------------[AppCompatImageView] id:org.kiwix.kiwixmobile:id/item_book_icon contDesc:Favicon visibility:visible[0m
2023-12-21T15:10:59.0335290Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: --------------[MaterialTextView] id:org.kiwix.kiwixmobile:id/item_book_title text:Test ZIM file visibility:visible[0m
2023-12-21T15:10:59.0337730Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: --------------[MaterialTextView] id:org.kiwix.kiwixmobile:id/item_book_description text:N/A visibility:visible[0m
2023-12-21T15:10:59.0339980Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: --------------[MaterialTextView] id:org.kiwix.kiwixmobile:id/item_book_date text:2022-11-18 visibility:visible[0m
2023-12-21T15:10:59.0342190Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: --------------[MaterialTextView] id:org.kiwix.kiwixmobile:id/item_book_size text:3 KB visibility:visible[0m
2023-12-21T15:10:59.0344730Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: --------------[MaterialTextView] id:org.kiwix.kiwixmobile:id/item_book_article_count text:1 articles visibility:visible[0m
2023-12-21T15:10:59.0347150Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: --------------[TagsView] id:org.kiwix.kiwixmobile:id/tags visibility:visible[0m
2023-12-21T15:10:59.0367490Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: ---------------[Chip] id:org.kiwix.kiwixmobile:id/tag_picture text:Pic visibility:visible[0m
2023-12-21T15:10:59.0369460Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: ---------------[Chip] id:org.kiwix.kiwixmobile:id/tag_video text:Vid visibility:visible[0m
2023-12-21T15:10:59.0371490Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: ---------------[Chip] id:org.kiwix.kiwixmobile:id/tag_short_text text:Short Text visibility:gone[0m
2023-12-21T15:10:59.0373470Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: ---------------[Chip] id:org.kiwix.kiwixmobile:id/tag_text_only text:Text Only visibility:gone[0m
2023-12-21T15:10:59.0375480Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: --------------[View] id:org.kiwix.kiwixmobile:id/item_book_clickable_area visibility:visible[0m
2023-12-21T15:10:59.0377230Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: ------------[MaterialCardView] visibility:visible[0m
2023-12-21T15:10:59.0378770Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: -------------[ConstraintLayout] visibility:visible[0m
2023-12-21T15:10:59.0380550Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: --------------[Guideline] id:org.kiwix.kiwixmobile:id/horizontal_padding visibility:gone[0m
2023-12-21T15:10:59.0382500Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: --------------[Guideline] id:org.kiwix.kiwixmobile:id/vertical_padding visibility:gone[0m
2023-12-21T15:10:59.0384500Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: --------------[MaterialCheckBox] id:org.kiwix.kiwixmobile:id/itemBookCheckbox visibility:visible[0m
2023-12-21T15:10:59.0386620Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: --------------[AppCompatImageView] id:org.kiwix.kiwixmobile:id/item_book_icon contDesc:Favicon visibility:visible[0m
2023-12-21T15:10:59.0389910Z [31m12-21 15:10:58.270 28792 28865 E TestRunner: --------------[MaterialTextView] id:org.kiwix.kiwixmobile:id/item_book_title text:Test_Zim visibility:visible[0m
2023-12-21T15:10:59.0392740Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: --------------[MaterialTextView] id:org.kiwix.kiwixmobile:id/item_book_description text:Wikipedia article on Test Zim visibility:visible[0m
2023-12-21T15:10:59.0395650Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: --------------[MaterialTextView] id:org.kiwix.kiwixmobile:id/item_book_date text:2017-04-21 visibility:visible[0m
2023-12-21T15:10:59.0398220Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: --------------[MaterialTextView] id:org.kiwix.kiwixmobile:id/item_book_size text:348 KB visibility:visible[0m
2023-12-21T15:10:59.0400500Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: --------------[MaterialTextView] id:org.kiwix.kiwixmobile:id/item_book_article_count text:4 articles visibility:visible[0m
2023-12-21T15:10:59.0402550Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: --------------[TagsView] id:org.kiwix.kiwixmobile:id/tags visibility:visible[0m
2023-12-21T15:10:59.0404660Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: ---------------[Chip] id:org.kiwix.kiwixmobile:id/tag_picture text:Pic visibility:visible[0m
2023-12-21T15:10:59.0406890Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: ---------------[Chip] id:org.kiwix.kiwixmobile:id/tag_video text:Vid visibility:visible[0m
2023-12-21T15:10:59.0408910Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: ---------------[Chip] id:org.kiwix.kiwixmobile:id/tag_short_text text:Short Text visibility:gone[0m
2023-12-21T15:10:59.0410930Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: ---------------[Chip] id:org.kiwix.kiwixmobile:id/tag_text_only text:Text Only visibility:gone[0m
2023-12-21T15:10:59.0412930Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: --------------[View] id:org.kiwix.kiwixmobile:id/item_book_clickable_area visibility:visible[0m
2023-12-21T15:10:59.0415170Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: -----------[MaterialButton] id:org.kiwix.kiwixmobile:id/startServerButton text:Start server visibility:visible[0m
2023-12-21T15:10:59.0417300Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: --------[BottomNavigationView] id:org.kiwix.kiwixmobile:id/bottom_nav_view visibility:gone[0m
2023-12-21T15:10:59.0419120Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: ---------[BottomNavigationMenuView] visibility:visible[0m
2023-12-21T15:10:59.0421130Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: ----------[BottomNavigationItemView] id:org.kiwix.kiwixmobile:id/readerFragment contDesc:Reader visibility:visible[0m
2023-12-21T15:10:59.0423390Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: -----------[AppCompatImageView] id:org.kiwix.kiwixmobile:id/navigation_bar_item_icon_view visibility:visible[0m
2023-12-21T15:10:59.0425600Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: -----------[BaselineLayout] id:org.kiwix.kiwixmobile:id/navigation_bar_item_labels_group visibility:visible[0m
2023-12-21T15:10:59.0435050Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: ------------[MaterialTextView] id:org.kiwix.kiwixmobile:id/navigation_bar_item_small_label_view text:Reader visibility:visible[0m
2023-12-21T15:10:59.0445120Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: ------------[MaterialTextView] id:org.kiwix.kiwixmobile:id/navigation_bar_item_large_label_view text:Reader visibility:invisible[0m
2023-12-21T15:10:59.0463530Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: ----------[BottomNavigationItemView] id:org.kiwix.kiwixmobile:id/libraryFragment contDesc:Library visibility:visible[0m
2023-12-21T15:10:59.0465940Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: -----------[AppCompatImageView] id:org.kiwix.kiwixmobile:id/navigation_bar_item_icon_view visibility:visible[0m
2023-12-21T15:10:59.0468160Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: -----------[BaselineLayout] id:org.kiwix.kiwixmobile:id/navigation_bar_item_labels_group visibility:visible[0m
2023-12-21T15:10:59.0470480Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: ------------[MaterialTextView] id:org.kiwix.kiwixmobile:id/navigation_bar_item_small_label_view text:Library visibility:invisible[0m
2023-12-21T15:10:59.0473800Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: ------------[MaterialTextView] id:org.kiwix.kiwixmobile:id/navigation_bar_item_large_label_view text:Library visibility:visible[0m
2023-12-21T15:10:59.0476460Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: ----------[BottomNavigationItemView] id:org.kiwix.kiwixmobile:id/downloadsFragment contDesc:Download visibility:visible[0m
2023-12-21T15:10:59.0479080Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: -----------[AppCompatImageView] id:org.kiwix.kiwixmobile:id/navigation_bar_item_icon_view visibility:visible[0m
2023-12-21T15:10:59.0481370Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: -----------[BaselineLayout] id:org.kiwix.kiwixmobile:id/navigation_bar_item_labels_group visibility:visible[0m
2023-12-21T15:10:59.0483690Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: ------------[MaterialTextView] id:org.kiwix.kiwixmobile:id/navigation_bar_item_small_label_view text:Download visibility:visible[0m
2023-12-21T15:10:59.0486150Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: ------------[MaterialTextView] id:org.kiwix.kiwixmobile:id/navigation_bar_item_large_label_view text:Download visibility:invisible[0m
2023-12-21T15:10:59.0488530Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: -------[NavigationView] id:org.kiwix.kiwixmobile:id/drawer_nav_view visibility:invisible[0m
2023-12-21T15:10:59.0490630Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: --------[NavigationMenuView] id:org.kiwix.kiwixmobile:id/design_navigation_view visibility:visible[0m
2023-12-21T15:10:59.0493150Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: ---------[LinearLayout] id:org.kiwix.kiwixmobile:id/navigation_header_container visibility:visible[0m
2023-12-21T15:10:59.0495070Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: ----------[LinearLayout] visibility:visible[0m
2023-12-21T15:10:59.0496320Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: -----------[AppCompatImageView] visibility:visible[0m
2023-12-21T15:10:59.0497890Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: ---------[NavigationMenuItemView] id:org.kiwix.kiwixmobile:id/menu_bookmarks_list visibility:visible[0m
2023-12-21T15:10:59.0500390Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: ----------[AppCompatCheckedTextView] id:org.kiwix.kiwixmobile:id/design_menu_item_text text:Bookmarks visibility:visible[0m
2023-12-21T15:10:59.0502620Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: ----------[ViewStub] id:org.kiwix.kiwixmobile:id/design_menu_item_action_area_stub visibility:gone[0m
2023-12-21T15:10:59.0504820Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: ---------[NavigationMenuItemView] id:org.kiwix.kiwixmobile:id/menu_history visibility:visible[0m
2023-12-21T15:10:59.0507050Z [31m12-21 15:10:58.271 28792 28865 E TestRunner: ----------[AppCompatCheckedTextView] id:org.kiwix.kiwixmobile:id/design_menu_item_text text:History visibility:visible[0m
2023-12-21T15:10:59.0509260Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: ----------[ViewStub] id:org.kiwix.kiwixmobile:id/design_menu_item_action_area_stub visibility:gone[0m
2023-12-21T15:10:59.0511320Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: ---------[NavigationMenuItemView] id:org.kiwix.kiwixmobile:id/menu_notes visibility:visible[0m
2023-12-21T15:10:59.0514280Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: ----------[AppCompatCheckedTextView] id:org.kiwix.kiwixmobile:id/design_menu_item_text text:Notes visibility:visible[0m
2023-12-21T15:10:59.0516500Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: ----------[ViewStub] id:org.kiwix.kiwixmobile:id/design_menu_item_action_area_stub visibility:gone[0m
2023-12-21T15:10:59.0518570Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: ---------[NavigationMenuItemView] id:org.kiwix.kiwixmobile:id/menu_host_books visibility:visible[0m
2023-12-21T15:10:59.0520510Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: ----------[AppCompatCheckedTextView] id:org.kiwix.kiwixmobile:id/design_menu_item_text text:WiFi Hotspot visibility:visible[0m
2023-12-21T15:10:59.0523380Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: ----------[ViewStub] id:org.kiwix.kiwixmobile:id/design_menu_item_action_area_stub visibility:gone[0m
2023-12-21T15:10:59.0525470Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: ---------[FrameLayout] visibility:visible[0m
2023-12-21T15:10:59.0527110Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: ----------[View] visibility:visible[0m
2023-12-21T15:10:59.0529080Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: ---------[NavigationMenuItemView] id:org.kiwix.kiwixmobile:id/menu_settings visibility:visible[0m
2023-12-21T15:10:59.0531680Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: ----------[AppCompatCheckedTextView] id:org.kiwix.kiwixmobile:id/design_menu_item_text text:Settings visibility:visible[0m
2023-12-21T15:10:59.0533960Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: ----------[ViewStub] id:org.kiwix.kiwixmobile:id/design_menu_item_action_area_stub visibility:gone[0m
2023-12-21T15:10:59.0535760Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: ---------[FrameLayout] visibility:visible[0m
2023-12-21T15:10:59.0537210Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: ----------[View] visibility:visible[0m
2023-12-21T15:10:59.0538650Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: ---------[NavigationMenuItemView] id:org.kiwix.kiwixmobile:id/menu_help visibility:visible[0m
2023-12-21T15:10:59.0540520Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: ----------[AppCompatCheckedTextView] id:org.kiwix.kiwixmobile:id/design_menu_item_text text:Help visibility:visible[0m
2023-12-21T15:10:59.0659130Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: ----------[ViewStub] id:org.kiwix.kiwixmobile:id/design_menu_item_action_area_stub visibility:gone[0m
2023-12-21T15:10:59.0660940Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: ---------[NavigationMenuItemView] id:org.kiwix.kiwixmobile:id/menu_support_kiwix visibility:visible[0m
2023-12-21T15:10:59.0662900Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: ----------[AppCompatCheckedTextView] id:org.kiwix.kiwixmobile:id/design_menu_item_text text:Support Kiwix visibility:visible[0m
2023-12-21T15:10:59.0664790Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: ----------[ViewStub] id:org.kiwix.kiwixmobile:id/design_menu_item_action_area_stub visibility:gone[0m
2023-12-21T15:10:59.0666620Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: -------[NavigationView] id:org.kiwix.kiwixmobile:id/reader_drawer_nav_view visibility:invisible[0m
2023-12-21T15:10:59.0743360Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: --------[NavigationMenuView] id:org.kiwix.kiwixmobile:id/design_navigation_view visibility:visible[0m
2023-12-21T15:10:59.0745850Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: ---------[LinearLayout] id:org.kiwix.kiwixmobile:id/navigation_header_container visibility:visible[0m
2023-12-21T15:10:59.0748150Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: ----------[RecyclerView] id:org.kiwix.kiwixmobile:id/right_drawer_list visibility:visible[0m
2023-12-21T15:10:59.0750330Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: -----[ViewStubCompat] id:org.kiwix.kiwixmobile:id/action_mode_bar_stub visibility:gone[0m
2023-12-21T15:10:59.0752390Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: --[View] id:android:id/navigationBarBackground visibility:visible[0m
2023-12-21T15:10:59.0754240Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: --[View] id:android:id/statusBarBackground visibility:visible[0m
2023-12-21T15:10:59.0755600Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: [0m
2023-12-21T15:10:59.0757150Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: 	at DebugFunctionsKt$applyWithViewHierarchyPrinting$1$1.run(DebugFunctions.kt:45)[0m
2023-12-21T15:10:59.0759210Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:463)[0m
2023-12-21T15:10:59.0761050Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: 	at java.util.concurrent.FutureTask.run(FutureTask.java:264)[0m
2023-12-21T15:10:59.0763740Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: 	at android.app.Instrumentation$SyncRunnable.run(Instrumentation.java:2379)[0m
2023-12-21T15:10:59.0765550Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: 	at android.os.Handler.handleCallback(Handler.java:942)[0m
2023-12-21T15:10:59.0767820Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: 	at android.os.Handler.dispatchMessage(Handler.java:99)[0m
2023-12-21T15:10:59.0769250Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: 	at android.os.Looper.loopOnce(Looper.java:201)[0m
2023-12-21T15:10:59.0770750Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: 	at android.os.Looper.loop(Looper.java:288)[0m
2023-12-21T15:10:59.0772100Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: 	at android.app.ActivityThread.main(ActivityThread.java:7872)[0m
2023-12-21T15:10:59.0773740Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: 	at java.lang.reflect.Method.invoke(Native Method)[0m
2023-12-21T15:10:59.0775560Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)[0m
2023-12-21T15:10:59.0777450Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)[0m
2023-12-21T15:10:59.0779360Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: Caused by: java.lang.RuntimeException: No view found with text WiFi connection detected[0m
2023-12-21T15:10:59.0781230Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: 	at org.kiwix.kiwixmobile.BaseRobot.isVisible(BaseRobot.kt:57)[0m
2023-12-21T15:10:59.0783040Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: 	at org.kiwix.kiwixmobile.BaseRobot.isVisible$default(BaseRobot.kt:56)[0m
2023-12-21T15:10:59.0784970Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: 	at org.kiwix.kiwixmobile.webserver.ZimHostRobot.startServer(ZimHostRobot.kt:73)[0m
2023-12-21T15:10:59.0787830Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: 	at org.kiwix.kiwixmobile.webserver.ZimHostFragmentTest$testZimHostFragment$2.invoke(ZimHostFragmentTest.kt:118)[0m
2023-12-21T15:10:59.0790040Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: 	at org.kiwix.kiwixmobile.webserver.ZimHostFragmentTest$testZimHostFragment$2.invoke(ZimHostFragmentTest.kt:104)[0m
2023-12-21T15:10:59.0792590Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: 	at org.kiwix.kiwixmobile.webserver.ZimHostRobotKt.zimHost(ZimHostRobot.kt:157)[0m
2023-12-21T15:10:59.0794860Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: 	at org.kiwix.kiwixmobile.webserver.ZimHostFragmentTest.testZimHostFragment(ZimHostFragmentTest.kt:104)[0m
2023-12-21T15:10:59.0796820Z [31m12-21 15:10:58.272 28792 28865 E TestRunner: 	at java.lang.reflect.Method.invoke(Native Method)[0m
2023-12-21T15:10:59.0799060Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)[0m
2023-12-21T15:10:59.0801150Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)[0m
2023-12-21T15:10:59.0803210Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)[0m
2023-12-21T15:10:59.0805260Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)[0m
2023-12-21T15:10:59.0807370Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at androidx.test.internal.runner.junit4.statement.RunBefores.evaluate(RunBefores.java:80)[0m
2023-12-21T15:10:59.0809480Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at androidx.test.internal.runner.junit4.statement.RunAfters.evaluate(RunAfters.java:61)[0m
2023-12-21T15:10:59.0811690Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at androidx.test.rule.GrantPermissionRule$RequestPermissionStatement.evaluate(GrantPermissionRule.java:136)[0m
2023-12-21T15:10:59.0814410Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at org.kiwix.kiwixmobile.testutils.RetryRule$statement$1.evaluate(RetryRule.kt:38)[0m
2023-12-21T15:10:59.0816460Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)[0m
2023-12-21T15:10:59.0818760Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)[0m
2023-12-21T15:10:59.0820950Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)[0m
2023-12-21T15:10:59.0822930Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)[0m
2023-12-21T15:10:59.0824970Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)[0m
2023-12-21T15:10:59.0826850Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)[0m
2023-12-21T15:10:59.0828670Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)[0m
2023-12-21T15:10:59.0830460Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)[0m
2023-12-21T15:10:59.0832250Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)[0m
2023-12-21T15:10:59.0835160Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)[0m
2023-12-21T15:10:59.0839420Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)[0m
2023-12-21T15:10:59.0843830Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)[0m
2023-12-21T15:10:59.0846100Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at org.junit.runners.Suite.runChild(Suite.java:128)[0m
2023-12-21T15:10:59.0848010Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at org.junit.runners.Suite.runChild(Suite.java:27)[0m
2023-12-21T15:10:59.0850470Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)[0m
2023-12-21T15:10:59.0852540Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)[0m
2023-12-21T15:10:59.0855110Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)[0m
2023-12-21T15:10:59.0857390Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)[0m
2023-12-21T15:10:59.0859760Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)[0m
2023-12-21T15:10:59.0862230Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)[0m
2023-12-21T15:10:59.0863710Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)[0m
2023-12-21T15:10:59.0865070Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)[0m
2023-12-21T15:10:59.0866380Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)[0m
2023-12-21T15:10:59.0867870Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at androidx.test.internal.runner.TestExecutor.execute(TestExecutor.java:67)[0m
2023-12-21T15:10:59.0869490Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at androidx.test.internal.runner.TestExecutor.execute(TestExecutor.java:58)[0m
2023-12-21T15:10:59.0871140Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at androidx.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:446)[0m
2023-12-21T15:10:59.0873250Z [31m12-21 15:10:58.273 28792 28865 E TestRunner: 	at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:2361)[0m
2023-12-21T15:10:59.0874880Z [31m12-21 15:10:58.274 28792 28865 E TestRunner: ----- end exception -----[0m
2023-12-21T15:10:59.0876450Z [32m12-21 15:10:58.297 28792 28865 I TestRunner: finished: testZimHostFragment(org.kiwix.kiwixmobile.webserver.ZimHostFragmentTest)[0m
2023-12-21T15:10:59.0878440Z [32m12-21 15:10:58.310 28792 28792 I MonitoringInstr: Finishing activity: org.kiwix.kiwixmobile.main.KiwixMainActivity@55ea121[0m
2023-12-21T15:10:59.2747290Z 
2023-12-21T15:10:59.2767570Z org.kiwix.kiwixmobile.webserver.ZimHostFragmentTest > testZimHostFragment[test(AVD) - 13] [31mFAILED [0m
2023-12-21T15:10:59.2770180Z 	java.lang.RuntimeException: No view found with text WiFi connection detected
```

```kotlin
SaveSearchToRecentsTest > invoke with non null Id saves search() FAILED
    java.lang.AssertionError at SaveSearchToRecentsTest.kt:62
```

--- a/app/src/androidTest/java/org/kiwix/kiwixmobile/webserver/ZimHostFragmentTest.kt
+++ b/app/src/androidTest/java/org/kiwix/kiwixmobile/webserver/ZimHostFragmentTest.kt
@@ -53,14 +53,16 @@ class ZimHostFragmentTest {
     arrayOf(
       Manifest.permission.READ_EXTERNAL_STORAGE,
       Manifest.permission.WRITE_EXTERNAL_STORAGE,
-      Manifest.permission.NEARBY_WIFI_DEVICES
+      Manifest.permission.NEARBY_WIFI_DEVICES,
+      Manifest.permission.ACCESS_NETWORK_STATE
     )
   } else {
     arrayOf(
       Manifest.permission.READ_EXTERNAL_STORAGE,
       Manifest.permission.WRITE_EXTERNAL_STORAGE,
       Manifest.permission.ACCESS_COARSE_LOCATION,
-      Manifest.permission.ACCESS_FINE_LOCATION
+      Manifest.permission.ACCESS_FINE_LOCATION,
+      Manifest.permission.ACCESS_NETWORK_STATE
     )
   }
 

--- a/app/src/androidTest/java/org/kiwix/kiwixmobile/webserver/ZimHostRobot.kt
+++ b/app/src/androidTest/java/org/kiwix/kiwixmobile/webserver/ZimHostRobot.kt
@@ -24,6 +24,7 @@ import androidx.test.espresso.action.ViewActions.click
 import androidx.test.espresso.assertion.ViewAssertions.matches
 import androidx.test.espresso.matcher.ViewMatchers
 import androidx.test.espresso.matcher.ViewMatchers.assertThat
+import androidx.test.espresso.matcher.ViewMatchers.withText
 import applyWithViewHierarchyPrinting
 import com.adevinta.android.barista.interaction.BaristaSleepInteractions
 import com.adevinta.android.barista.interaction.BaristaSwipeRefreshInteractions.refresh
@@ -64,14 +65,21 @@ class ZimHostRobot : BaseRobot() {
   }
 
   fun clickOnTestZim() {
+    pauseForBetterTestPerformance()
     clickOn(Text("Test_Zim"))
   }
 
   fun startServer() {
+    // stop the server if it is already running.
+    stopServerIfAlreadyStarted()
     clickOn(ViewId(R.id.startServerButton))
+    assetWifiDialogDisplayed()
+    onView(withText("PROCEED")).perform(click())
+  }
+
+  private fun assetWifiDialogDisplayed() {
     pauseForBetterTestPerformance()
-    isVisible(TextId(R.string.wifi_dialog_title))
-    clickOn(TextId(R.string.hotspot_dialog_neutral_button))
+    isVisible(Text("WiFi connection detected"))
   }
 
   fun assertServerStarted() {
@@ -110,7 +118,6 @@ class ZimHostRobot : BaseRobot() {
   }
 
   private fun selectZimFile(position: Int) {
-    pauseForBetterTestPerformance()
     try {
       onView(
         RecyclerViewMatcher(R.id.recyclerViewZimHost).atPositionOnView(
@@ -119,7 +126,6 @@ class ZimHostRobot : BaseRobot() {
         )
       ).check(matches(ViewMatchers.isChecked()))
     } catch (assertionError: AssertionFailedError) {
-      pauseForBetterTestPerformance()
       onView(
         RecyclerViewMatcher(R.id.recyclerViewZimHost).atPositionOnView(
           position,
@@ -148,6 +154,6 @@ class ZimHostRobot : BaseRobot() {
   }
 
   private fun pauseForBetterTestPerformance() {
-    BaristaSleepInteractions.sleep(TestUtils.TEST_PAUSE_MS_FOR_DOWNLOAD_TEST.toLong())
+    BaristaSleepInteractions.sleep(TestUtils.TEST_PAUSE_MS.toLong())
   }
 }

--- a/contrib/instrumentation.sh
+++ b/contrib/instrumentation.sh
@@ -1,5 +1,7 @@
 #!/usr/bin/env bash
 
+# Enable Wi-Fi on the emulator
+adb shell svc wifi enable
 adb logcat -c
 adb logcat ./*:E -v color &
 retry=0
@@ -11,6 +13,8 @@ do
   else
     adb kill-server
     adb start-server
+    # Enable Wi-Fi on the emulator
+    adb shell svc wifi enable
     adb logcat -c
     adb logcat ./*:E -v color &
     ./gradlew clean

--- a/contrib/instrumentation.sh
+++ b/contrib/instrumentation.sh
@@ -3,7 +3,8 @@
 # Enable Wi-Fi on the emulator
 adb shell svc wifi enable
 adb logcat -c
-adb logcat ./*:E -v color &
+# shellcheck disable=SC2035
+adb logcat *:E -v color &
 retry=0
 while [ $retry -le 3 ]
 do
@@ -16,7 +17,8 @@ do
     # Enable Wi-Fi on the emulator
     adb shell svc wifi enable
     adb logcat -c
-    adb logcat ./*:E -v color &
+    # shellcheck disable=SC2035
+    adb logcat *:E -v color &
     ./gradlew clean
     retry=$(( retry + 1 ))
     if [ $retry -eq 3 ]; then

--- a/core/src/test/java/org/kiwix/kiwixmobile/core/search/viewmodel/effects/SaveSearchToRecentsTest.kt
+++ b/core/src/test/java/org/kiwix/kiwixmobile/core/search/viewmodel/effects/SaveSearchToRecentsTest.kt
@@ -23,9 +23,9 @@ import io.mockk.Called
 import io.mockk.mockk
 import io.mockk.verify
 import kotlinx.coroutines.ExperimentalCoroutinesApi
-import kotlinx.coroutines.test.TestCoroutineScope
+import kotlinx.coroutines.test.TestScope
 import kotlinx.coroutines.test.advanceUntilIdle
-import kotlinx.coroutines.test.runBlockingTest
+import kotlinx.coroutines.test.runTest
 import org.junit.jupiter.api.Test
 import org.kiwix.kiwixmobile.core.dao.NewRecentSearchDao
 import org.kiwix.kiwixmobile.core.reader.ZimFileReader
@@ -38,7 +38,7 @@ internal class SaveSearchToRecentsTest {
   private val searchListItem = RecentSearchListItem("", ZimFileReader.CONTENT_PREFIX)
 
   private val activity: AppCompatActivity = mockk()
-  private val testDispatcher = TestCoroutineScope()
+  private val testDispatcher = TestScope()
 
   @Test
   fun `invoke with null Id does nothing`() {
@@ -54,7 +54,7 @@ internal class SaveSearchToRecentsTest {
   }
 
   @Test
-  fun `invoke with non null Id saves search`() = testDispatcher.runBlockingTest {
+  fun `invoke with non null Id saves search`() = testDispatcher.runTest {
     val id = "id"
     SaveSearchToRecents(
       newRecentSearchDao,

--- a/core/src/test/java/org/kiwix/kiwixmobile/core/search/viewmodel/effects/SaveSearchToRecentsTest.kt
+++ b/core/src/test/java/org/kiwix/kiwixmobile/core/search/viewmodel/effects/SaveSearchToRecentsTest.kt
@@ -22,20 +22,23 @@ import androidx.appcompat.app.AppCompatActivity
 import io.mockk.Called
 import io.mockk.mockk
 import io.mockk.verify
-import kotlinx.coroutines.CoroutineScope
-import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.test.TestCoroutineScope
+import kotlinx.coroutines.test.advanceUntilIdle
+import kotlinx.coroutines.test.runBlockingTest
 import org.junit.jupiter.api.Test
 import org.kiwix.kiwixmobile.core.dao.NewRecentSearchDao
 import org.kiwix.kiwixmobile.core.reader.ZimFileReader
 import org.kiwix.kiwixmobile.core.search.adapter.SearchListItem.RecentSearchListItem
 
+@OptIn(ExperimentalCoroutinesApi::class)
 internal class SaveSearchToRecentsTest {
 
   private val newRecentSearchDao: NewRecentSearchDao = mockk()
   private val searchListItem = RecentSearchListItem("", ZimFileReader.CONTENT_PREFIX)
 
   private val activity: AppCompatActivity = mockk()
-  private val viewModelScope = CoroutineScope(Dispatchers.IO)
+  private val testDispatcher = TestCoroutineScope()
 
   @Test
   fun `invoke with null Id does nothing`() {
@@ -43,7 +46,7 @@ internal class SaveSearchToRecentsTest {
       newRecentSearchDao,
       searchListItem,
       null,
-      viewModelScope
+      testDispatcher
     ).invokeWith(
       activity
     )
@@ -51,14 +54,15 @@ internal class SaveSearchToRecentsTest {
   }
 
   @Test
-  fun `invoke with non null Id saves search`() {
+  fun `invoke with non null Id saves search`() = testDispatcher.runBlockingTest {
     val id = "id"
     SaveSearchToRecents(
       newRecentSearchDao,
       searchListItem,
       id,
-      viewModelScope
+      testDispatcher
     ).invokeWith(activity)
+    testDispatcher.advanceUntilIdle()
     verify { newRecentSearchDao.saveSearch(searchListItem.value, id, ZimFileReader.CONTENT_PREFIX) }
   }
 }

